text/0097-log-data-model.md

```diff
@@ -0,0 +1,971 @@
+14            |WARN2
+15            |WARN3
+16            |WARN4
+17            |ERROR
+18            |ERROR2
+19            |ERROR3
+20            |ERROR4
+21            |FATAL
+22            |FATAL2
+23            |FATAL3
+24            |FATAL4
+
+When an individual log record is displayed it is recommended to show both
+`SeverityText` and `SeverityNumber` values. A recommended combined string in
+this case begins with the short name followed by `SeverityText` in parenthesis.
+
+For example "Informational" Syslog record will be displayed as **INFO
+(Informational)**. When for a particular log record the `SeverityNumber` is
+defined but the `SeverityText` is missing it is recommended to only show the
+short name, e.g. **INFO**.
+
+When drop down lists (or other UI elements that are intended to represent the
+possible set of values) are used for representing the severity it is preferable
+to display the short name in such UI elements.
+
+For example a dropdown list of severities that allows filtering log records by
+severities is likely to be more usable if it contains the short names of
+`SeverityNumber` (and thus has a limited upper bound of elements) compared to a
+dropdown list, which lists all distinct `SeverityText` values that are known to
+the system (which can be a large number of elements, often differing only in
+capitalization or abbreviated, e.g. "Info" vs "Information").
+
+#### Comparing Severity
+
+In the contexts where severity participates less-than / greater-than comparisons
+`SeverityNumber` field should be used. `SeverityNumber` can be compared to
+another `SeverityNumber` or to numbers in the 1..24 range (or to the
+corresponding short names).
+
+When severity is used in equality or inequality comparisons (for example in
+filters in the UIs) the recommendation is to attempt to use both `SeverityText`
+and short name of `SeverityNumber` to perform matches. For example if we have a
+record with `SeverityText` field equal to "Informational" and `SeverityNumber`
+field equal to INFO then it may be preferable from the user experience
+perspective to ensure that **severity="Informational"** and **severity="INFO"**
+conditions both to are TRUE for that record.
+
+### Field: `ShortName`
+
+Type: string.
+
+Description: Short event identifier that does not contain varying parts.
+`ShortName` describes what happened (e.g. "ProcessStarted"). Recommended to be
+no longer than 50 characters. Optional. Not guaranteed to be unique in any way.
+Typically used for filtering and grouping purposes in backends.
+
+### Field: `Body`
+
+Type: any.
+
+Description: A value containing the body of the log record (see the description
+of `any` type above). Can be for example a human-readable string message
+(including multi-line) describing the event in a free form or it can be a
+structured data composed of arrays and maps of other values. Can vary for each
+occurrence of the event coming from the same source.
+
+### Field: `Resource`
```

Comment:
The characterization of a resource can deserves 15+ fields. Using domain namespaces as grouping elements rather than having a flat list would help the readability and thus help the implementation. 

These fields being very useful to quickly correlate observability data on any dimension (component version, host, data center...).

Here is an example with the “donut_shop” application deployed as an Helm chart on Kubernetes on AWS to illustrate how long the list of fields can be.


```
├── resource
│   ├── service
│   │   ├── name: donut-shop
│   │   ├── feature_flags: ["anti_fraud_2020_04_01", "cross_selling_2020_04_12"]
│   │   └── version: 2.0.0
│   ├── helm_chart
│   │   └── name: donut-shop
│   │   └── app_version: 2.0.0
│   │   └── version: 2.0.0
│   ├── container
│   │   ├── image.name: docker.mycie.com/donut_shop
│   │   └── image.tag: 2.0.0
│   ├── k8s
│   │   └── cluster.name: website-production
│   │   └── deployment.name: production-donut-shop
│   │   └── pod.name: production-donut-shop-7758b9cdf8-cs798
│   │   └── service.name: production-donut-shop
│   │   └── service.external_ip:  # <none>
│   ├── host
│   │   └── id: i-1234567890abcdef0
│   │   └── ip: 10.101.163.102
│   ├── cloud
│   │   └── zone: us-east-1c
│   │   └── region: us-east-1
│   ├── labels
│   │   └── my.custom.application.tag: hello
```

Reply:
@cyrille-leclerc we do have a "domain/namespace" in the flattened map. I think adding concrete levels will just add complexity and more important overhead on the wire - need to create a lot of maps which are not free when deserialize. Some examples here:
https://github.com/open-telemetry/opentelemetry-specification/tree/master/specification/resource/semantic_conventions

Reply:
Thanks @bogdandrutu, I used the field definitions of https://github.com/open-telemetry/opentelemetry-specification/tree/master/specification/resource/semantic_conventions

I feel that grouping fields by their namespace helps the readability. I acknowledge that there is a subjective dimension in the readability.

Here is the flat representation of the same set of fields:

```
├── resource
|   ├── service.name: donut-shop
|   ├── service.feature_flags: ["anti_fraud_2020_04_01", "cross_selling_2020_04_12"]
|   ├── service.version: 2.0.0
|   ├── helm_chart.name: donut-shop
|   ├── helm_chart.app_version: 2.0.0
|   ├── helm_chart.version: 2.0.0
|   ├── container.image.name: docker.mycie.com/donut_shop
|   ├── container.image.tag: 2.0.0
|   ├── k8s.cluster.name: website-production
|   ├── k8s.deployment.name: production-donut-shop
|   ├── k8s.pod.name: production-donut-shop-7758b9cdf8-cs798
|   ├── k8s.service.name: production-donut-shop
|   ├── k8s.service.external_ip:  # <none>
|   ├── host.id: i-1234567890abcdef0
|   ├── host.ip: 10.101.163.102
|   ├── cloud.zone: us-east-1c
|   ├── cloud.region: us-east-1
|   ├── labels.my.custom.application.tag: hello
```

Reply:
I do see the difference, but I think this is just a visualization concern and not a protocol concern, what do you think?

Reply:
I would tend to agree with you that we can probably decouple visualization and protocol / storage / transport.

I am not knowledgeable enough on document/column oriented data structures to understand the pros on cons of the representation with "list & dots" of hierarchical data structures.

Reply:
@cyrille-leclerc I agree with you that hierarchical representation is more readable (subjectively, I know :-) ).

However, the definition of the Resource in this proposal is consistent with how it is used for Traces and Metrics in OpenTelemetry. This is simply re-iteration of what already is part of the Traces and Metrics spec. I believe it is highly important to keep the consistency between 3 signals unless there are very good reasons to diverge.

The dot notation that is currently used by OpenTelemetry spec is more or less equivalent to the hierarchical representation (except some edge cases like having dots in attribute names). I agree with Bogdan on the following:
- The dot notation can be visualized as a hierarchy, essentially matching the look of ECS approach.
- Flattened map may be more efficient on the wire (needs benchmarking).

I am not against any proposals to improve Resource representation, but I believe it needs to be done uniformly for all telemetry signals not just logs. That requires a separate proposal/OTEP. If that separate proposal gets accepted I will be happy to modify this data model accordingly.

Reply:
Recorded 2 issues to work on Resource representation improvements:
https://github.com/open-telemetry/opentelemetry-specification/issues/579
https://github.com/open-telemetry/opentelemetry-specification/issues/583

Resolving this comment.